src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -904,6 +903,105 @@ internal RunVersionOne CreateRun(Run v2Run)
+        // v2 run.files array to find the required FileData object, we construct a
+        // dictionary from the array so we can access the required FileData object
+        // directly. It turns out that we need mappings in both directions: from
+        // array index to dictionary key, and vice versa.
+        private static void CreateFileKeyIndexMappings(
+            IList<FileData> v2Files,
+            out IDictionary<string, int> fileKeyToIndexDictionary,
+            out IDictionary<int, string> fileIndexToKeyDictionary)
+        {
+            fileKeyToIndexDictionary = null;
+            fileIndexToKeyDictionary = null;
+
+            if (v2Files != null)
+            {
+                fileKeyToIndexDictionary = new Dictionary<string, int>();
+                fileIndexToKeyDictionary = new Dictionary<int, string>();
+
+                for (int index = 0; index < v2Files.Count; ++index)
+                {
+                    FileData v2File = v2Files[index];
+                    string key = CreateFileDictionaryKey(v2File, v2Files);
+
+                    fileKeyToIndexDictionary[key] = index;
+                    fileIndexToKeyDictionary[index] = key;
+                }
+            }
+        }
+
+        // Given a v2 FileData object, synthesize a key that will be used to locate that object
+        // in the v2 file data dictionary. We will use the same key to locate the corresponding
+        // v1 FileData object when we create the v1 run.files dictionary.
+        //
+        // This method is necessary because, although ideally we would use the FileData
+        // object's URI as the dictionary key, it is possible for two FileData objects to
+        // have the same URI. For example:
+        //
+        //    1. Nested files with the same name might exist in two different containers.
+        //    2. There might be two files with the same URI but difference uriBaseIds.
+        //
+        // To deal with Case #1, we synthesize a key as follows:
+        //
+        //    <root file URI>#<nested file path>/<level 2 nested file path>/...
+        //
+        // To deal with Case #2, we would follow the chain or uriBaseIds, prepending
+        // the value of each one to the URI.
+        //
+        // WE DO NOT YET HANDLE CASE #2.
+        private static string CreateFileDictionaryKey(FileData v2File, IList<FileData> v2Files)
+        {
+            var sb = new StringBuilder(v2File.FileLocation.Uri.OriginalString);
+            while (v2File.ParentIndex != -1)
+            {
+                FileData parentFile = v2Files[v2File.ParentIndex];
+
+                // The convention for building the key is as follows:
+                // The root file URI is separated from the chain of nested files by '#';
+                // The nested file URIs are separated from each other by '/'.
+                string separator = parentFile.ParentIndex == -1 ? "#" : "/";
```

Comment:
>string separator = parentFile.ParentIndex [](start = 16, length = 41)

are you sure this code doesn't double slashes at the point where we jump from a nested archive to an element in it? 

i.e., need a test case for myarchive.zip#myarchive2.zip/foo.cpp

in this three-part file object, the most nested item will have a URL of /foo.cpp

does your code double that slash? like so myarchive.zip#myarchive2.zip//foo.cpp?

Reply:
Good catch! I forgot that nested file URIs already start with a slash. I'll add the test case you asked for to expose this bug, and then fix it.

---
In reply to: [247376946](https://github.com/Microsoft/sarif-sdk/pull/1211#discussion_r247376946) [](ancestors = 247376946)

Reply:
Fixed.

---
In reply to: [247570514](https://github.com/Microsoft/sarif-sdk/pull/1211#discussion_r247570514) [](ancestors = 247570514,247376946)